src/rsa/padding.rs

```diff
@@ -150,3 +150,138 @@ impl Verification for PKCS1 {
+/// Salt length is fixed to be the length of the output of the digest algorithm.
+/// The only mask generating function supported is MGF1.
+/// The size (in bits) of the RSA moduli is assumed to be a multiple of 8.
+pub struct PSS {
+    pub digest_alg: &'static digest::Algorithm,
+}
+
+// Maxmimum supported output length for PSS padding.
+const MAX_OUTPUT_LEN: usize = 4096;
+
+impl Verification for PSS {
+    // PSS verification as specified in
+    // https://tools.ietf.org/html/rfc3447#section-9.1.2
+    fn verify(&self, msg: untrusted::Input, encoded: untrusted::Input)
+              -> Result<(), ()> {
+
+        let em_len = encoded.len();
+        encoded.read_all((), |em| {
+            let digest_len = self.digest_alg.output_len;
+
+            // Step 2.
+            let m_hash = digest::digest(self.digest_alg,
+                                        msg.as_slice_less_safe());
+
+            // Step 3: where we assume the digest and salt are of equal length.
+            if em_len < 2 + (2 * digest_len) {
+                return Err(())
+            }
+
+            // Steps 4 and 5: Parse encoded message as:
+            //     masked_db || h_hash || 0xbc
+            let db_len = em_len - digest_len - 1;
+            let masked_db = try!(em.skip_and_get_input(db_len));
+            let h_hash = try!(em.skip_and_get_input(digest_len));
+            if try!(em.read_byte()) != 0xbc {
+                return Err(());
+            }
+
+            // Step 7.
+            let mut db = [0u8; MAX_OUTPUT_LEN];
+            try!(mgf1(h_hash.as_slice_less_safe(), db_len, &mut db,
+                      self.digest_alg));
+
+            try!(masked_db.read_all((), |masked_bytes| {
+                // Step 6. Moduli lengths are always a multiple of 8 so the top
+                // bit must be 0.
+                let b = try!(masked_bytes.read_byte());
+                if b & 0x80 != 0 {
+                    return Err(());
+                } else {
+                    db[0] ^= b;
+                }
+
+                // Step 8.
+                for i in 1..db_len {
+                    let b = try!(masked_bytes.read_byte());
+                    db[i] ^= b;
+                }
+                Ok(())
+            }));
+
+            // Step 9.
+            db[0] &= 0x7f;
+
+            // Step 10.
+            let pad_len = db_len - digest_len - 1;
+            for i in 0..pad_len {
+                if db[i] != 0 {
+                    return Err(());
+                }
+            }
+            if db[pad_len] != 1 {
+                return Err(());
+            }
+
+            // Step 11.
+            let salt = &db[db_len - digest_len..][..digest_len];
+
+            // Step 12 and 13: compute hash value of:
+            //     (0x)00 00 00 00 00 00 00 00 || m_hash || salt
+            let mut ctx = digest::Context::new(self.digest_alg);
+            let prefix = [0u8; 8];
```

Comment:
this should be `const PSS_PREFIX_ZEROS: [u8; 8] = [0u8; 8]` and should be shared between the encoding and decoding functions. And/or, maybe this whole block from `let mut ctx...ctx.finish()` should be factored out into a shared function?